virtcontainers/pkg/types/pcipath.go

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) 2020 David Gibson, Red Hat Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+package types
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+)
+
+const maxPciSlot = 31
+
+// A PciPath describes where a PCI sits in a PCI hierarchy.
+//
+// Consists of a list of PCI slot numbers, giving the slot of each
+// bridge that must be traversed from the PCI root to reach the
+// device, followed by the slot of the device itself
+//
+// When formatted into a string is written as "xx/.../yy/zz" Here, zz
+// is the slot of the device on its PCI bridge, yy is the slot of the
+// bridge on its parent bridge and so forth until xx is the slot of
+// the "most upstream" bridge on the root bus.  If a device is
+// connected directly to the root bus, its PciPath is just "zz"
+
+type PciPath struct {
+	path []uint8
+}
+
+func (p PciPath) Format() string {
+	tokens := make([]string, len(p.path))
+	for i, slot := range p.path {
+		tokens[i] = fmt.Sprintf("%02x", slot)
+	}
+	return strings.Join(tokens, "/")
+}
+
+func (p PciPath) IsNil() bool {
+	return p.path == nil
+}
+
+func PciPathFromString(s string) (PciPath, error) {
+	if s == "" {
+		return PciPath{}, nil
+	}
+
+	tokens := strings.Split(s, "/")
+	slots := make([]uint8, len(tokens))
+	for i, t := range tokens {
+		slot, err := strconv.ParseUint(t, 16, 5)
+		if err != nil {
+			return PciPath{}, err
+		}
+		slots[i] = uint8(slot)
+	}
+	return PciPath{path: slots}, nil
+}
+
+func MakePciPath(slots ...uint8) (PciPath, error) {
```

Comment:
Why is this variadic? It seems we only ever pass two values...?

Reply:
For now, but I have future plans that will change that.  With the agent side changes, it will now accept and work with paths of an arbitrary number of elements.

Reply:
Hmm. Conventionally PRs shouldn't do this as you are spreading a feature across multiple PRs which, is cause for potential confusion. Why not just make the function accept 2 params and then the follow-on PR (when written) can change this function to be a variadic since that change is part of the future work?

Reply:
Hmm.  (a) that seems unnecessarily bureaucratic.  (b) I don't think the division you propose is actually conceptually sounder than this one.

What this series is doing - and the first patch in particular - is introducing a data structure which will handle general pci paths to replace the ad-hoc and limited form of pci paths we deal with now.  At present we only use one special case of pci paths, but it doesn't seem sensible to artificially limit the data structure just because we don't need the general variant yet.